README.md
Geocod.io PHP
============

[![Build Status](https://travis-ci.org/davidstanley01/geocodio-php.png?branch=master)](https://travis-ci.org/davidstanley01/geocodio-php)

Thin PHP wrapper for [Geocod.io geocoding API](http://geocod.io/docs) that includes a service provider for Laravel.

# Features

* Geocode an individual address
* Batch geocode up to 10,000 addresses at a time
* Parse an address into its identifiable components

Read the complete [Geocod.io geocoding API](http://geocod.io/docs) for service documentation.

# Installing via Composer

The recommended way to install Geocod.io PHP is through [Composer](http://getcomposer.org).

```bash
# Install Composer
curl -sS https://getcomposer.org/installer | php

# Add Geocodio as a dependency
php composer.phar require stanley/geocodio-php:~1.0
```

After installing, you need to require Composer's autoloader:

```php
require 'vendor/autoload.php';
```

Using
-----

Using the Geocod.io PHP library is super simple.
```php
require_once 'vendor/autoload.php'
use Stanley\Geocodio\Client;

// Create the new Client object by passing in your api key
$client = new Client('YOUR_API_KEY');
```

## Single address
To encode a single address, use the `get` method and pass in your address as a string.
```php
$data = '123 Anywhere St, Chicago, IL';
$address = $client->get($data);
```

## Multiple addresses
For multiple addresses, call the 'post' method and pass in your data as either an object or an array.
```php
$data = [
  '123 Anywhere St, Chicago, IL',
  '456 Oak St, Los Angeles, CA'
];
$address = $client->post($data);
```
You can also call the `$client->geocode($data)` method with either a single address as a `string` or mulitple addresses as an `array` or an `object` and the method will call the appropriate endpoint.

## Parsing a single address
To parse an address, call the 'parse' method and pass in your data as a string
```php
$data = '123 Anywhere St, Chicago, IL';
$address = $client->parse($data);
```

## Shortcut
To make things even simpler, you can also call the `$client->geocode($data)` method with either a single address as a `string` or mulitple addresses as an `array` or an `object` and the method will call the appropriate endpoint.

## Reverse Geocoding
Geocod.io also supports reverse geocoding.  To convert a single lat/long pair into an address, call the 'reverse' method.  The lat/long pairs should be a comma-separated string.

```php
$data = "35.9746000,-77.9658000";
$address = $client->reverse($data);
```

To do a bulk conversion, pass your lat/long pairs in an array.
```php
$data = ["35.9746000,-77.9658000","32.8793700,-96.6303900","33.8337100,-117.8362320","35.4171240,-80.6784760"];
$address = $client->reverse($data);
```

## Return Values
The Geocod.io PHP client will return an instance of the `Stanley\Geocodio\Data` class.  The raw response body can be accessed from the `response` property. This property had the response data stored as an object.

```php
$body = $address->response;
```





Soon, I'll add some helper methods so you can easily pick certain elements out.  Coming soon...

## Laravel Service Provider and Facade
A service provder and facade are available if you are using Laravel.  Once you've added the package to your composer.json file, run `php composer.phar update`. Add `'Stanley\Geocodio\ServiceProviders\LaravelServiceProvider'`to the `providers` key.  Then, edit the `aliases` key and add `'Geocodio' => 'Stanley\Geocodio\Geocodio'` to the array.

You are now ready to use the Facade.  If you are using Laravel, you will need to pass your API Key as the second parameter.
```php
$key = 'YOUR_API_KEY';
$data = Geocodio::get('123 Anywhere St, Chicago, IL', $key);
return Response::json($data);
```

Exceptions
----------

Periodically, the Geocod.io service will return errors.

    200 OK Hopefully you will see this most of the time. Note that this status code will also be returned even though no geocoding results were available
    403 Forbidden Invalid API key or other reason why access is forbidden
    422 Unprocessable Entity A client error prevented the request from executing succesfully (e.g. invalid address provided). A JSON object will be returned with an error key containing a full error message
    500 Server Error Hopefully you will never see this...it means that something went wrong in our end. Whoops.

To handle these:

* An HTTP 403 error raises a `GeocodioAuthError`
* An HTTP 422 error raises a `GeocodioDataError` and the error message will be
  reported through the exception
* An HTTP 5xx error raises a `GeocodioServerError`
* An unmatched non-200 response will simply raise `Exception`

Credits
-------
Much of this readme and the structure of this library was inspired by the [Py-Geocodio Library](https://github.com/bennylope/pygeocodio) by [Ben Lopatin](https://github.com/bennylope).

HISTORY.md
History
=======

1.1.0 (2014-02-03)
------------------
* Added reverse geocoding endpoint.

1.0.6 (2014-01-26)
------------------
* Added integration tests and fixed issue where post data was being encoded twice.

1.0.5 (2014-01-23)
------------------
* Specified minimum version of Guzzle at >=3.7.0

1.0.4 (2014-01-22)
------------------
* Added service provider and facade for Laravel
* Added method to auto-detect endpoint required for single or multiple address (not for `parse` endpoint)
* Refactored some methods to allow for API Key to be set after instantiation via setter or optional parameter.
* Added GeocodioException instead of using general exception

1.0.3 (2014-01-22)
------------------
* Forgot to bump version in composer.json

1.0.2 (2014-01-22)
------------------
* Fixed mis-named class property

1.0.1 (2014-01-22)
------------------
* Added license and changelog

1.0.0 (2014-01-22)
------------------

* Initial release to Packagist

src/Stanley/Geocodio/Client.php
<?php namespace Stanley\Geocodio;

use Stanley\Geocodio\Data;
use Guzzle\Http\Client as GC;
use Guzzle\Http\Message\Response;

class Client
{
    const BASE_URL = 'http://api.geocod.io/v1/';

    /**
     * API Key
     * @var string
     */
    protected $apiKey;

    /**
     * Guzzle Client object
     * @var Guzzle\Http\Client
     */
    protected $client;

    /**
     * Class constructor
     *
     * @param string $apiKey API Key
     */
    public function __construct($apiKey = null)
    {
        $this->apiKey = $apiKey;
        $this->client = $this->newGuzzleClient();
    }

    /**
     * Setter for API Key
     *
     * @param string $apiKey API Key
     * @return void
     */
    public function setApiKey($apiKey)
    {
        $this->apiKey = $apiKey;
    }

    /**
     * Geocode facade
     *
     * @param  mixed $data incoming data
     * @param  string $key  API Key
     * @return mixed
     */
    public function geocode($data, $key = null)
    {
        if ($key) $this->apiKey = $key;
        return (is_string($data)) ? $this->get($data) : $this->post($data);
    }
    /**
     * Get Method
     *
     * @param  string $data Data to be encoded
     * @param  string $verb URL segment to call - either 'geocode' or 'parse'
     * @return Stanley\Geocodio\Data
     */
    public function get($data, $key = null, $verb = 'geocode')
    {
        if ($key) $this->apiKey = $key;
        $request = $this->getRequest($data, $verb);
        return $this->newDataObject($request->getBody());
    }

    /**
     * Post Method
     *
     * @param  array $data Data to be encoded
     * @return Stanley\Geocodio\Data
     */
    public function post($data, $key = null)
    {
        if ($key) $this->apiKey = $key;
        $request = $this->bulkPost($data);
        return $this->newDataObject($request->getBody());
    }

    /**
     * Parse method
     *
     * @param  string $data Data to be encoded
     * @return Stanley\Geocodio\Data
     */
    public function parse($data, $key = null)
    {
        if ($key) $this->apiKey = $key;
        return $this->get($data, null, 'parse');
    }

    /**
     * Reverse Method
     *
     * @param  mixed  $data Information to encode
     * @param  string $key  API Key
     * @return Stanley\Geocodio\Data
     */
    public function reverse($data, $key = null)
    {
        return (is_string($data)) ? $this->get($data, $key, 'reverse') : $this->post($data, $key, 'reverse');
    }

    /**
     * Call Guzzle with Get Request
     *
     * @param  string $data Address Data
     * @param  string $verb The method being called - geocode, parse, or reverse
     * @return Guzzle\Http\Message\Response
     */
    protected function getRequest($data, $verb)
    {
        $params = [
            'q' => str_replace(' ', '+', $data),
            'api_key' => $this->apiKey
        ];
        $request = $this->client->get(self::BASE_URL . $verb, [], [
            'query' => $params
        ]);
        $response = $this->client->send($request);
        return $this->checkResponse($response);
    }

    /**
     * Call Guzzle Post request method
     *
     * @param  array $data Address data
     * @param  string $verb Method to be called - should only be geocode for now
     * @return Guzzle\Http\Message\Response
     */
    protected function bulkPost($data, $verb = 'geocode')
    {
        $url = self::BASE_URL . $verb . "?api_key=" . $this->apiKey;
        $headers = [ 'Content-Type' => 'application/json' ];
        $payload = json_encode($data);

        $request = $this->client->post($url, $headers, $payload, []);
        $response = $this->client->send($request);
        return $this->checkResponse($response);
    }

    /**
     * Check response code and throw appropriate exception
     *
     * @param  Guzzle\Http\Message\Respone $response Guzzle Response
     * @return mixed
     */
    protected function checkResponse(Response $response)
    {
        $status = $response->getStatusCode();
        $reason = $response->getReasonPhrase();
        switch ($status) {
            case '403':
                throw new Stanley\Geocodio\GeocodioAuthError($reason);
                break;

            case '422':
                throw new Stanley\Geocodio\GeocodioDataError($reason);
                break;

            case '500':
                throw new Stanley\Geocodio\GeocodioServerError($reason);
                break;

            case '200':
                return $response;
                break;

            default:
                throw new Stanley\Geocodio\GeocodioException("There was a problem with your request - $reason");
                break;
        }
    }

    /**
     * Create new Guzzle Client
     *
     * @return Guzzle\Http\Client
     */
    protected function newGuzzleClient()
    {
        return new GC();
    }

    /**
     * Create new Data object
     *
     * @param  mixed $response Response body
     * @return Stanley\Geocodio\Data
     */
    protected function newDataObject($response)
    {
        return new Data($response);
    }
}